### PR TITLE
[SPARK-36069][SQL] Add field info to `from_json`'s exception in the FAILFAST mode

### PR DESCRIPTION
**Pull request by geekyouth**
### What changes were proposed in this pull request?

spark function from_json output field name, field type and field value when FAILFAST mode throw exception.

### Why are the changes needed?

This infoormation is very important for devlops to find where error input data is located.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

org/apache/spark/sql/JsonFunctionsSuite.scala:598
test("[SPARK-36069] from_json invalid json schema - check field name and field value")

### CODE REVIEW

**Comment by geekyouth** [comment]
@HyukjinKwon Look here 

Another one PR can be close:
https://github.com/apache/spark/pull/33292

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
WDYT @MaxGekk?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -595,6 +595,31 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+      val df = Seq(badRec, """{"a": 2, "b": 12}""").toDS()
+
+      checkAnswer(
+        df.select(from_json($"value", schema, Map("mode" -> "PERMISSIVE"))),
+        Row(Row(null, 11, badRec)) :: Row(Row(2, 12, null)) :: Nil)
+
+      val exception = intercept[SparkException] {
+        df.select(from_json($"value", schema, Map("mode" -> "FAILFAST"))).collect()
+      }
+      exception.printStackTrace()
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:614]: Could you remove this, please.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 2]
```diff
@@ -595,6 +595,31 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+      assert(exception.getMessage.contains(
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:615]: Store `exception.getMessage` in a val, and reuse it in both asserts. For example:
```scala
      val errMsg = intercept[SparkException] {
        df.select(from_json($"value", schema, Map("mode" -> "FAILFAST"))).collect()
      }. getMessage
```

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 3]
```diff
@@ -595,6 +595,31 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("[SPARK-36069] from_json invalid json schema - check field name and field value") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:598]:
```suggestion
  test("SPARK-36069: from_json invalid json schema - check field name and field value") {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 4]
```diff
@@ -958,18 +958,25 @@ object QueryExecutionErrors {
     new RuntimeException("Parsing JSON arrays as structs is forbidden.")
   }
 
-  def cannotParseStringAsDataTypeError(str: String, dataType: DataType): Throwable = {
-    new RuntimeException(s"Cannot parse $str as ${dataType.catalogString}.")
+  def cannotParseStringAsDataTypeError(parser: JsonParser, token: JsonToken, dataType: DataType)
+  : Throwable = {
+    new RuntimeException(
+      s"Cannot parse field name [${parser.getCurrentName}], " +
+        s"field value [${parser.getText}], " +
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:965]: Any reason to output field name and values in such form? Could you check other places in Spark code base, and look for other forms of field names and values in error messages.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 5]
```diff
@@ -958,18 +958,25 @@ object QueryExecutionErrors {
...
   }
 
-  def failToParseValueForDataTypeError(dataType: DataType, token: JsonToken): Throwable = {
+  def failToParseValueForDataTypeError(parser: JsonParser, token: JsonToken, dataType: DataType)
+  : Throwable = {
     new RuntimeException(
-      s"Failed to parse a value for data type ${dataType.catalogString} (current token: $token).")
+      s"Failed to parse field name [${parser.getCurrentName}], " +
+        s"field value [${parser.getText}], " +
+        s"[${token.toString}] to target spark data type [${dataType}].")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:979]:
```suggestion
        s"[$token] to target spark data type [$dataType].")
```

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @geekyouth .

**Comment by geekyouth** [comment]
> +1, LGTM. Merging to master.
> Thank you, @geekyouth .

thanks, but why this pr closed instead of merged?😳

**Comment by geekyouth** [comment]
I also want to merge this feature into version 2.4.

Because version 3.0+ is not compatible with ES and other compoment.

And this feature is very important for the spark app which running in PROD ENV.
@MaxGekk 
Do you think it is acceptable?

**Pull request closed without merging**